src/rules/noUnsafeFinallyRule.ts

```diff
@@ -40,171 +39,111 @@ export class Rule extends Lint.Rules.AbstractRule {
-        return this.applyWithWalker(new NoReturnInFinallyScopeAwareWalker(sourceFile, this.getOptions()));
+        return this.applyWithFunction(sourceFile, walk);
     }
 }
 
-/**
- * Represents details associated with tracking finally scope.
- */
-interface IFinallyScope {
-    /**
-     * A value indicating whether the current scope is a `break` boundary.
-     */
-    isBreakBoundary: boolean;
-
-    /**
-     * A value indicating whether the current scope is a `continue` boundary.
-     */
-    isContinueBoundary: boolean;
-
-    /**
-     * A value indicating whether the current scope is within a finally block.
-     */
-    isFinallyBlock: boolean;
-
-    /**
-     * A value indication whether the current scope is a `return` and `throw` boundary.
-     */
-    isReturnsThrowsBoundary: boolean;
-
-    /**
-     * A collection of `break` or `continue` labels in this scope.
-     */
-    labels: string[];
-}
-
-/**
- * Represents a block walker that identifies finally blocks and walks
- * only the blocks that do not change scope for return statements.
- */
-class NoReturnInFinallyScopeAwareWalker extends Lint.ScopeAwareRuleWalker<IFinallyScope> {
-
-    protected visitBreakStatement(node: ts.BreakOrContinueStatement) {
-        if (!this.isControlFlowWithinFinallyBlock(isBreakBoundary, node)) {
-            super.visitBreakStatement(node);
-            return;
-        }
-
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_BREAK));
-    }
-
-    protected visitContinueStatement(node: ts.BreakOrContinueStatement) {
-        if (!this.isControlFlowWithinFinallyBlock(isContinueBoundary, node)) {
-            super.visitContinueStatement(node);
-            return;
-        }
-
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_CONTINUE));
-    }
-
-    protected visitLabeledStatement(node: ts.LabeledStatement) {
-        this.getCurrentScope().labels.push(node.label.text);
-
-        super.visitLabeledStatement(node);
-    }
-
-    protected visitReturnStatement(node: ts.ReturnStatement): void {
-        if (!this.isControlFlowWithinFinallyBlock(isReturnsOrThrowsBoundary)) {
-            super.visitReturnStatement(node);
-            return;
-        }
-
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_RETURN));
-    }
-
-    protected visitThrowStatement(node: ts.ThrowStatement): void {
-        if (!this.isControlFlowWithinFinallyBlock(isReturnsOrThrowsBoundary)) {
-            super.visitThrowStatement(node);
-            return;
+function walk(ctx: Lint.WalkContext<void>): void {
+    let inFinally = false;
+    ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        switch (node.kind) {
+            case ts.SyntaxKind.TryStatement:
+                const { tryBlock, catchClause, finallyBlock } = node as ts.TryStatement;
+                ts.forEachChild(tryBlock, cb);
+                if (catchClause) {
```

Comment:
yeah, I'd like to enable it, I remember having some issues when trying to use it elsewhere but I haven't put in the time to investigate it fully.